### ISSUE TITLE
Amplify.configure() hangs on hot restart when using AmplifyPushNotificationsPinpoint

### ISSUE THREAD

**Issue by sbkomurcu** [aws-amplify/amplify-flutter#2989]
### Description

I have been following the official documentation to configure both AmplifyAuthCognito and AmplifyPushNotificationsPinpoint for my Flutter app. However, I have encountered an issue where configuring both libraries simultaneously causes the Authenticator to become stuck on the loading screen. 

Interestingly, I discovered that when I comment out the line for AmplifyPushNotificationsPinpoint, the app runs without any issues.
 
Tested only on Android.

I have included the relevant code snippet below for reference:
```
import 'package:flutter/material.dart';
import 'package:amplify_analytics_pinpoint/amplify_analytics_pinpoint.dart';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_authenticator/amplify_authenticator.dart';
import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:amplify_push_notifications_pinpoint/amplify_push_notifications_pinpoint.dart';


import 'amplifyconfiguration.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  AmplifyLogger().logLevel = LogLevel.info;
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
    _configureAmplify();
  }

  void _configureAmplify() async {
    try {
      await Amplify.addPlugins([
        AmplifyAnalyticsPinpoint(),
        AmplifyAuthCognito(),

        // *** Commenting out the following line will cause the app to work as expected. ***
        //  AmplifyPushNotificationsPinpoint()
      ]);
      await Amplify.configure(amplifyconfig);
    } on AmplifyAlreadyConfiguredException {
      safePrint(
        'Tried to reconfigure Amplify; this can occur when your app restarts on Android.',
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Authenticator(
      child: MaterialApp(
        builder: Authenticator.builder(),
        home: Scaffold(
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                const Text('You are logged in!'),
                ElevatedButton(
                  onPressed: () async {
                    await Amplify.Auth.signOut();
                  },
                  child: const Text('Sign Out'),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

```

### Categories

- [X] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [X] Authenticator
- [ ] DataStore
- [ ] Storage

### Steps to Reproduce

1. Create a Flutter app and integrate Amplify-Flutter for both authentication and push notifications.
2. Copy the code snippet provided above and paste it into the main.dart file.
3. Run the app, and it should work without any issues.
4. Close the app.
5. Comment out the line of code that configures AmplifyPushNotificationsPinpoint.
6. Run the app again, and it gets stuck showing the loading spinner.

### Screenshots

![bug](https://user-images.githubusercontent.com/19965011/236159554-9c67ec42-dd02-4b19-9638-97a82b47ba76.jpg)

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.7.12

### Amplify Flutter Version

1.0.1

### Deployment Method

Amplify CLI

**Comment by Jordan-Nelson** [comment]
@sbkomurcu - if the Authenticator is showing a loading indicator indefinitely that indicates that configuration never completed successfully. My guess is there was a configuration error.

I see you are catching AmplifyAlreadyConfiguredException. Can you try changing that to Exception and see if there is any exception thrown? Can you also confirm that the call to amplify.configure() is completing? You can temporarily add a print statement immediately after that line and see if it is being reached.

**Comment by sbkomurcu** [comment]
@Jordan-Nelson - Before creating this issue, I attempted to confirm that the configuration process had completed successfully. However, it appears to be stuck on the line `amplify.configure()` and is not producing any error messages. I have attached two images below to provide further context. 

**With AmplifyPushNotificationsPinpoint:**

![image](https://user-images.githubusercontent.com/19965011/236193281-ea357492-e3bb-4e45-b9f3-c46349d03239.png)

**Without AmplifyPushNotificationsPinpoint:**

![image](https://user-images.githubusercontent.com/19965011/236194650-08188610-eb72-49e4-87f9-7c3db1696441.png)

**Comment by Jordan-Nelson** [comment]
Okay. It seems like configuration is not completing when the push notification plugin is added. If you change the log level to verbose do you see any additional log output?

Can you share the contents of your amplify config? You can remove any app specific info or API keys.

**Comment by jamilsaadeh97** [comment]
I do have a similar problem whenever I hot restart my application. It get stuck on Amplify.configure() but without any exception being thrown.

I want to note that I am not using the Authenticator.

**Comment by sbkomurcu** [comment]
@Jordan-Nelson -  I can confirm that the same issue persists even when the `AmplifyAnalyticsPinpoint()` is not used at all. 

Here is the amplify config:

```
const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "analytics": {
        "plugins": {
            "awsPinpointAnalyticsPlugin": {
                "pinpointAnalytics": {
                    "appId": "************",
                    "region": "eu-central-1"
                },
                "pinpointTargeting": {
                    "region": "eu-central-1"
                }
            }
        }
    },
    "notifications": {
        "plugins": {
            "awsPinpointPushNotificationsPlugin": {
                "appId": "************",
                "region": "eu-central-1"
            }
        }
    },
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "API1": {
                    "endpointType": "GraphQL",
                    "endpoint": "************",
                    "region": "eu-central-1",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS",
                    "apiKey": "************",
                },
                "API2": {
                    "endpointType": "REST",
                    "endpoint": "************",
                    "region": "eu-central-1",
                    "authorizationType": "AWS_IAM"
                },
                "API3": {
                    "endpointType": "REST",
                    "endpoint":"************",
                    "region": "eu-central-1",
                    "authorizationType": "AWS_IAM"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "************",
                        "Region": "eu-central-1",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "************",
                    },
                    "************_API_KEY": {
                        "ApiUrl": "************",
                        "Region": "eu-central-1",
                        "AuthMode": "API_KEY",
                        "ApiKey": "************",
                        "ClientDatabasePrefix": "************",
                    },
                    "************_AWS_IAM": {
                        "ApiUrl": "************",
                        "Region": "eu-central-1",
                        "AuthMode": "AWS_IAM",
                        "ClientDatabasePrefix": "************",
                    }
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "************",
                            "Region": "eu-central-1"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "************",
                        "AppClientId": "************",
                        "Region": "eu-central-1"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": [],
                        "usernameAttributes": [
                            "EMAIL"
                        ],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": [
                                "REQUIRES_LOWERCASE",
                                "REQUIRES_UPPERCASE",
                                "REQUIRES_NUMBERS",
                                "REQUIRES_SYMBOLS"
                            ]
                        },
                        "mfaConfiguration": "OFF",
                        "mfaTypes": [],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                },
                "S3TransferUtility": {
                    "Default": {
                        "Bucket": "************",
                        "Region": "eu-central-1"
                    }
                },
                "PinpointAnalytics": {
                    "Default": {
                        "AppId": "************",
                        "Region": "eu-central-1"
                    }
                },
                "PinpointTargeting": {
                    "Default": {
                        "Region": "eu-central-1"
                    }
                }
            }
        }
    },
    "storage": {
        "plugins": {
            "awsS3StoragePlugin": {
                "bucket": "************",
                "region": "eu-central-1",
                "defaultAccessLevel": "guest"
            }
        }
    }
}''';
```

**Comment by Jordan-Nelson** [comment]
@jamilsaadeh97 - Is this specific to `AmplifyAnalyticsPinpoint()`? If not, can you open a new issue and fill out the info in the issue template?

**Comment by Jordan-Nelson** [comment]
@sbkomurcu - Have you followed the [platform specific setup for Android](https://docs.amplify.aws/lib/push-notifications/getting-started/q/platform/flutter/#prerequisites)? You will need to:
- Add `apply plugin: 'com.google.gms.google-services'` to the bottom of `android/app/build.gradle`
- Add `classpath 'com.google.gms:google-services:4.3.14'` to the list of dependencies in `android/build.gradle`
- Add your `google-services.json` file to `android/app`

I was able to reproduce a similar experience where `Amplify.configure()` never completed prior to taking those platform specific setup steps.

**Comment by jamilsaadeh97** [comment]
@Jordan-Nelson This is specific to `AmplifyPushNotificationsPinpoint()`. I can confirm after debugging it that hot restart is not working when I add the notifications plugin:
```dart
try {
    safePrint("Adding plugins");
    await Amplify.addPlugins([
      datastorePlugin,
      apiPlugin,
      authPlugin,
      analyticsPlugin,
      s3StoragePlugin,
      notificationsPlugin //If I comment it out, everything works fine.
    ]);
    safePrint("Plugins Added");
    
    safePrint("Configuring Amplify");
    await Amplify.configure(amplifyconfig);
    safePrint("Amplify Configured");
  } on AmplifyAlreadyConfiguredException {
    safePrint(
        'Tried to reconfigure Amplify; this can occur when your app restarts on Android.');
  } on Exception catch (error) {
    safePrint(error);
  }
```

If this issue is not related to the current one, I can open a new issue. Please let me know.

**Comment by Jordan-Nelson** [comment]
@jamilsaadeh97 Okay. I was able to reproduce that behavior. For now we can track these together as this may be the cause of the issue that @sbkomurcu is seeing.

@sbkomurcu - Are you seeing this issue on the initial app launch or only after hot restart?

**Comment by Jordan-Nelson** [comment]
@ALi4576 - if you are not using `AmplifyAnalyticsPinpoint()`, can you please open a new issue and fill out [the issue template](https://github.com/aws-amplify/amplify-flutter/issues/new?assignees=&labels=&template=BUG-REPORT.yaml)? This issue appears to be specific to `AmplifyAnalyticsPinpoint()`.

**Comment by ALi4576** [comment]
sure not problem. I am opening it.

**Comment by Jordan-Nelson** [comment]
@sbkomurcu - Please let me know if you are seeing this on initial app launch. I have only reproduced this on hot restart. I will mark this as a bug to have that behavior resolved.

**Comment by HuiSF** [comment]
Thank you Jordan for triaging!

>I have only reproduced this on hot restart.

This is due to that when Flutter hot restarts, the Android native implementation remains at the same state, and doesn't re-request device token, and this causes `AmplifyPushNotificationsPinpoint.configure` hang.

A workaround for this specific case is that you can minimize the App and reopen after hot restarted the Flutter App, this should unblock the `configure` call as the Android native implementation re-request device token on each App resume.

**Comment by sbkomurcu** [comment]
@Jordan-Nelson - Sorry for the late answer. 

> @sbkomurcu - Have you followed the [platform specific setup for Android](https://docs.amplify.aws/lib/push-notifications/getting-started/q/platform/flutter/#prerequisites)? You will need to:
> 
> I was able to reproduce a similar experience where `Amplify.configure()` never completed prior to taking those platform specific setup steps.

Following your suggestion, I have reconfigured the app using the `Legacy Plugin Application` instead of the previously used `Gradle plugins DSL`. I'm pleased to report that the app now launches as expected on the first launch.

> @sbkomurcu - Please let me know if you are seeing this on initial app launch. I have only reproduced this on hot restart. I will mark this as a bug to have that behavior resolved.

As previously mentioned, the app still experiences issues and becomes unresponsive after a hot-restart. Nonetheless, I can confirm that the workaround suggested by @HuiSF  is working for me as well.

**Comment by likeconan** [comment]
To help identify this issue, it happened in real device iOS 13.4.1

The error message is like below when I first flutter run:

```
[VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: InvalidAccountTypeException {
  "message": "No identity pool registered for this account",
  "recoverySuggestion": "Register an identity pool using the CLI"
}
#0      AWSResult.value (package:aws_common/src/result/aws_result.dart:54:9)
#1      CognitoIamAuthProvider.retrieve (package:amplify_auth_cognito_dart/src/util/cognito_iam_auth_provider.dart:26:42)
<asynchronous suspension>
#2      WithSigV4.intercept (package:smithy_aws/src/http/interceptors/with_sig_v4.dart:37:7)
<asynchronous suspension>
#3      SmithyHttpRequest.transformRequest (package:smithy/src/http/http_request.dart:137:19)
<asynchronous suspension>
#4      SmithyHttpRequest.send.<anonymous closure> (package:smithy/src/http/http_request.dart:188:23)
<asynchronous suspension>

```

Here is the amplify configuration file
```
{
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "anonymous-iam": {
                    "endpointType": "GraphQL",
                    "endpoint": "$graphqlEndpoint",
                    "region": "$region",
                    "authorizationType": "AWS_IAM"
                },
                "authenticated-with-user-pool": {
                    "endpointType": "GraphQL",
                    "endpoint": "$graphqlEndpoint",
                    "region": "$region",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "${AppConfig().account.poolId}",
                        "AppClientId": "${AppConfig().account.poolClientId}",
                        "Region": "$region"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": ["FACEBOOK","GOOGLE","APPLE"],
                        "usernameAttributes": ["EMAIL"],
                        "signupAttributes": [
                            "EMAIL"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 6,
                            "passwordPolicyCharacters": ["REQUIRES_UPPERCASE","REQUIRES_LOWERCASE","REQUIRES_NUMBERS"]
                        },
                        "mfaConfiguration": "OFF",
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                }
            }
        }
    },
    "notifications": {
      "plugins": {
        "awsPinpointPushNotificationsPlugin": {
          "appId": "***",
          "region": "$region"
        }
      }
    }
  }
```

And here is the configuration code:

```
    await Amplify.addPlugins([AmplifyAuthCognito(), AmplifyPushNotificationsPinpoint()]);
    await Amplify.configure(getAmplifyConfig());
```

When I hot restart, the error message is disappeared, but the configure call hangs.

Also, when I just remove AmplifyPushNotificationsPinpoint() plugin, and hot restart, everything works very well.

Please help to fix this issue, thanks

**Comment by HuiSF** [comment]
Hi @likeconan I don't think the exception you pasted related to this issue as it was thrown while getting credentials for signing requests.

From looking at your configuration it seems that you are not using the Amplify CLI generated file, if you are using any custom file, please make sure your code exactly matches the Amplify CLI generated content.  If you need further assistance please open a new issue.

**Comment by likeconan** [comment]
@HuiSF Hi, Yeah you are right, I need to setup cognito identity to make it work. My first issue can be ignored

**Comment by sbkomurcu** [comment]
Hi @Jordan-Nelson !  Is there an estimated timeline for the publication of the release?

**Comment by dnys1** [comment]
The fix for this has been released in v1.1.0 of the libraries (which now require Dart 3/Flutter 3.10).

Please let us know if you have a need for Dart 2/Flutter 3.7 support and we can backport it to Amplify v1.0.

**Comment by jamilsaadeh97** [comment]
Hi @dnys1 , I think that the issue is still there but in another form. 
I updated to the latest version and now when I hot restart, Amplify.configure() throws an `AmplifyAlreadyConfiguredException` which I catch. But the problem, in my case, is when  `await Amplify.Notifications.Push.getPermissionStatus();` got called, which is directly after Amplify.configure, it throws a `ConfigurationError ` exception: 
```
ConfigurationError (ConfigurationError {
  "message": "Configure Amplify with Notifications Plugin before using this method.",
  "recoverySuggestion": "Configure Amplify first."
})
```

**Comment by HuiSF** [comment]
Hi @jamilsaadeh97 I couldn't repro the issue you described.

Here's what I tested:

1. Click the hot restart button
2. Watched `Amplify.isConfigured` value as `false` so `Amplify.configure` ran normally and no `AmplifyAlreadyConfiguredException` was thrown
3.  calling `Amplify.Notifications.Push.getPermissionStatus()` right after `await Amplify.configure()`, it returned result correctly

**Comment by jamilsaadeh97** [comment]
Hi @HuiSF , did you try it on iOS or Android? I'm currently trying it on Android.

**Comment by HuiSF** [comment]
Both iOS and Android worked correctly.

**Comment by sprisao** [comment]
Thank you guys!
